File: src/triangvol.f90
subroutine trianvol(vb1,vb2,it,dimvb,dimit,V)
  implicit none
  integer :: i,j,dimvb,dimit,it(3,dimit),subvols(3,4)
  real*8 :: vb1(4,dimvb),vb2(4,dimvb),allvol(4,6), V,tmp,Vtmp

  subvols(1,:) = (/1,2,3,4/)
  subvols(2,:) = (/1,2,4,5/)
  subvols(3,:) = (/2,4,5,6/)
  V = 0.0d0
  do i = 1,dimit
     Vtmp = 0.0d0
     allvol(:,1:3) = vb1(:,it(:,i))
     allvol(:,(/5,6,4/)) = vb2(:,it(:,i))
     
     
     do j = 1,3
        call detfour(transpose(allvol(:,subvols(j,:))),tmp)
        Vtmp = Vtmp + abs(tmp)
     end do
     V = V + Vtmp/6

  end do
end subroutine trianvol

subroutine detfour(A,det)
 real*8:: A(4,4),det
DET =  A(1,1)*(A(2,2)*(A(3,3)*A(4,4)-A(3,4)*A(4,3))+A(2,3)*(A(3,4)*A(4,2)-A(3,2)*A(4,4))+A(2,4)*(A(3,2)*A(4,3)- &
             A(3,3)*A(4,2)))-A(1,2)*(A(2,1)*(A(3,3)*A(4,4)-A(3,4)*A(4,3))+A(2,3)*(A(3,4)*A(4,1)-A(3,1)*A(4,4))+ &
             A(2,4)*(A(3,1)*A(4,3)-A(3,3)*A(4,1)))+A(1,3)*(A(2,1)*(A(3,2)*A(4,4)-A(3,4)*A(4,2))+A(2,2)*(A(3,4)*A(4,1)- &
             A(3,1)*A(4,4))+A(2,4)*(A(3,1)*A(4,2)-A(3,2)*A(4,1)))-A(1,4)*(A(2,1)*(A(3,2)*A(4,3)-A(3,3)*A(4,2))+ &
             A(2,2)*(A(3,3)*A(4,1)-A(3,1)*A(4,3))+A(2,3)*(A(3,1)*A(4,2)-A(3,2)*A(4,1)))


end subroutine detfour

File: src/triangvol.cpp
#include "RcppArmadillo.h"

using namespace Rcpp;
using namespace arma;

typedef unsigned int uint;

RcppExport SEXP trianvol(SEXP vb1_, SEXP vb2_, SEXP it_) {
  double V = 0.0;
  
  try {
  mat armavb1 = as<mat>(vb1_);
  mat armavb2 = as<mat>(vb2_);
  umat uIT = as<umat>(it_);
  mat allvol(4,6);
  umat subvols(3,4);
  uvec tmp; tmp  << 0 << 1 << 2 << 3 << endr;
  subvols.row(0) = tmp.t();
  tmp  << 0 << 1 << 3 << 4 << endr;
  subvols.row(1) = tmp.t();
  tmp  <<  1 << 3 << 4 << 5 << endr;
  subvols.row(2) = tmp.t();
  uvec tmp02; tmp02 << 0 << 1 << 2 << endr;
  uvec tmp453; tmp453 << 4 << 5 << 3 << endr;
  int dimit = uIT.n_cols;
  for (unsigned int i = 0; i < dimit; i++) {
    double Vtmp = 0.0;
    uvec slice0 = uIT.col(i);
    allvol.cols(tmp02) = armavb1.cols(uIT.col(i));
    allvol.cols(tmp453) = armavb2.cols(uIT.col(i));
    for (unsigned int j = 0; j < 3; j++) {
      mat allvolsub = allvol.cols(subvols.row(j));
      allvolsub = allvolsub.t();
      double tmpdet = det(allvolsub);
      Vtmp += std::abs(tmpdet);
    }
    V += Vtmp/6;
  }
  } catch (...)
  {
  V=NA_REAL;
  }
  
  return wrap(V);
}
